webapp/src/components/SimRecorder.tsx

```diff
@@ -0,0 +1,187 @@
+import * as simulator from "../simulator";
+import * as React from "react"
+import * as screenshot from "../screenshot";
+import { getEditorAsync } from "../app";
+import { SimRecorder, SimRecorderRef, SimRecorderState } from "../../../react-common/components/share/ThumbnailRecorder";
+
+interface SimRecorderRefImpl extends SimRecorderRef {
+    gifAddFrame(data: ImageData, delay?: number): boolean;
+}
+
+export const SimRecorderImpl: SimRecorder = props => {
+    const { onSimRecorderInit } = props;
+
+    const [loanedSimulator, setLoanedSimulator] = React.useState<HTMLElement>(undefined);
+
+    React.useEffect(() => {
+        const sim = simulator.driver.loanSimulator()
+        setLoanedSimulator(sim);
+
+        const ref = createSimRecorderRef();
+
+        const onSimulatorMessage = (message: pxt.editor.ScreenshotData) => {
+            if ((message as any).type === "screenshot") {
+                if (ref.state === "recording") {
+                    // Adds frame, returns true if we've exceeded the max frame count
+                    if (ref.gifAddFrame(message.data, message.delay)) {
+                        ref.stopRecordingAsync();
+                    }
+                } else {
+                    ref.screenshotAsync();
+                }
+            } else if (message.event === "start") {
+                ref.startRecordingAsync();
+            } else if (message.event === "stop") {
+                ref.stopRecordingAsync();
+            }
+        }
+
+        getEditorAsync().then(editor => {
+            editor.pushScreenshotHandler(onSimulatorMessage)
+
+            onSimRecorderInit(ref);
+        })
+
+        return () => {
+            if (loanedSimulator) {
+                simulator.driver.unloanSimulator();
+                simulator.driver.stopRecording();
+            }
+            getEditorAsync().then(editor => {
+                editor.popScreenshotHandler()
+            })
+        }
+    }, [])
+
+    let containerRef: HTMLDivElement;
+    const handleContainerRef = (ref: HTMLDivElement) => {
+        if (!ref) return;
+
+        containerRef = ref;
+
+        if (loanedSimulator && containerRef.firstChild !== loanedSimulator) {
+            while (containerRef.firstChild) containerRef.firstChild.remove();
+            containerRef.appendChild(loanedSimulator);
+        }
+    }
+
+    return <div id="shareLoanedSimulator" ref={handleContainerRef}>
+    </div>
+}
+
+let ref: SimRecorderRefImpl;
+
+function createSimRecorderRef() {
+    if (ref) return ref;
+    let encoder: screenshot.GifEncoder;
+    ref = {
```

Comment:
nit; since ref declaration is hoisted we could just move this to the very bottom without the need for an any cast and then have all the anon functions attached in the object literal definition? Doesn't really matter though I suppose